lms/djangoapps/courseware/module_render.py

```diff
@@ -345,7 +381,52 @@ def inner_get_module(descriptor):
             request_token=request_token,
         )
 
-    def handle_grade_event(block, event_type, event):
+    def _fulfill_content_milestones(course_key, content_key, user_id):
+        """
+        Internal helper to handle milestone fulfillments for the specified content module
+        """
+        # If this module is part of an entrance exam, we'll need to see if the student
+        # has reached the point at which they can collect the associated milestone
+        if settings.FEATURES.get('ENTRANCE_EXAMS', False):
+            course = modulestore().get_course(course_key)
+            if getattr(course, 'entrance_exam_enabled', False):
+                exam_key = UsageKey.from_string(course.entrance_exam_id)
+                exam_descriptor = modulestore().get_item(exam_key)
+                exam_modules = yield_dynamic_descriptor_descendents(
+                    exam_descriptor,
+                    inner_get_module
+                )
+                ignore_categories = ['course', 'chapter', 'sequential', 'vertical']
+                module_pcts = []
+                exam_pct = 0
+                for module in exam_modules:
+                    if module.graded and module.category not in ignore_categories:
+                        module_pct = 0
+                        try:
+                            student_module = StudentModule.objects.get(
+                                module_state_key=module.scope_ids.usage_id
+                            )
+                            if student_module.max_grade:
+                                module_pct = student_module.grade / student_module.max_grade
+                        except StudentModule.DoesNotExist:
+                            pass
+                        module_pcts.append(module_pct)
+                exam_pct = sum(module_pcts) / float(len(module_pcts))
+                if exam_pct >= course.entrance_exam_minimum_score_pct:
+                    content_milestones = milestones_api.get_course_content_milestones(
+                        course_key,
+                        content_key,
+                        relationship='fulfills'
```

Comment:
Also, I'm pretty sure the Analytics team is going to want to have an analytics event fired here. You should check in with them. @brianhw can you - or someone else - weigh in on this?